pjrt-plugin/iree/integrations/pjrt/common/api_impl.cc

```diff
@@ -438,19 +457,83 @@ iree_status_t BufferInstance::GetHostSizeInBytes(iree_host_size_t* host_size) {
-  iree_hal_device_t* hal_device;
-  IREE_RETURN_IF_ERROR(device_.GetHalDevice(&hal_device));
-  IREE_RETURN_IF_ERROR(iree_hal_device_transfer_d2h(
-      hal_device, iree_hal_buffer_view_buffer(buffer_view()), 0, dst, dst_size,
-      IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT, iree_infinite_timeout()));
-
-  *done_event = new EventInstance();
+                                         EventInstance** out_done_event) {
+  // Set up an event for external triggering. While a little wonky, we
+  // trigger it in the host buffer release callback, which happens once the
+  // transfer is done. I don't love this option but it seems to match what
+  // I'm looking for.
+  EventInstance* capture_done_event =
+      new EventInstance(EventInstance::Type::EXTERNAL);
+  *out_done_event = capture_done_event;
+
+  // Import the destination (host) buffer as an iree_hal_buffer_t so that we
+  // can issue copy commands.
+  iree::vm::ref<iree_hal_buffer_t> dst_buffer;
+  iree_hal_buffer_params_t dst_buffer_params = {
+      IREE_HAL_BUFFER_USAGE_TRANSFER_TARGET};
+  iree_hal_external_buffer_t dst_external_buffer;
+  memset(&dst_external_buffer, 0, sizeof(dst_external_buffer));
+  dst_external_buffer.type = IREE_HAL_EXTERNAL_BUFFER_TYPE_HOST_ALLOCATION;
+  dst_external_buffer.flags = IREE_HAL_EXTERNAL_BUFFER_FLAG_NONE;
+  dst_external_buffer.size = dst_size;
+  dst_external_buffer.handle.host_allocation.ptr = dst;
+  auto release_callback = +[](void* user_data, iree_hal_buffer_t* buffer) {
+    IREE_TRACE_SCOPE0("PJRT_CopyToHost_ReleaseCallback");
+    auto* local_done_event = static_cast<EventInstance*>(user_data);
+    local_done_event->ExternalSignalReady(iree_ok_status());
+  };
+  IREE_RETURN_IF_ERROR(iree_hal_allocator_import_buffer(
+      device_.device_allocator(), dst_buffer_params, &dst_external_buffer,
+      /*release_callback=*/{release_callback, capture_done_event},
+      &dst_buffer));
+
+  // Create the transfer command buffer.
+  iree::vm::ref<iree_hal_command_buffer_t> transfer_cb;
+  iree_hal_transfer_command_t transfer_command;
+  memset(&transfer_command, 0, sizeof(transfer_command));
+  transfer_command.type = IREE_HAL_TRANSFER_COMMAND_TYPE_COPY;
+  transfer_command.copy.source_buffer =
+      iree_hal_buffer_view_buffer(buffer_view());
+  transfer_command.copy.source_offset = 0;
+  transfer_command.copy.target_buffer = dst_buffer.get();
+  transfer_command.copy.target_offset = 0;
+  transfer_command.copy.length = dst_size;
+  IREE_RETURN_IF_ERROR(iree_hal_create_transfer_command_buffer(
+      device_.device(), IREE_HAL_COMMAND_BUFFER_MODE_ONE_SHOT,
+      IREE_HAL_QUEUE_AFFINITY_ANY,
+      /*transfer_count=*/1, &transfer_command, &transfer_cb));
+  dst_buffer.reset();
+
+  IREE_RETURN_IF_ERROR(iree_hal_device_queue_execute(
+      device_.device(), IREE_HAL_QUEUE_AFFINITY_ANY,
+      /*wait_semaphore_list=*/iree_hal_fence_semaphore_list(ready_fence_.get()),
+      /*signal_semaphore_list=*/iree_hal_semaphore_list_empty(),
```

Comment:
RE comment above: you'll want a signal fence for this so when it's implemented you'll be able to hook on to that to get your event signal callback (or ideally if it's a standard event type like an fd/HANDLE import it so we can signal it directly from the GPU). You'll at least need the signal so that you ensure you don't schedule more work while this copy is outstanding: you need to make sure nothing is able to mutate the device buffer.